### ISSUE TITLE
Windows Server 2019 closes connection on swarm overlay network after ~6 Minutes

### ISSUE THREAD

**Issue by dmduggan** [moby/moby#44082]
### Description

After upgrading from Server 2016/Docker 18.09.14 to Server 2019/Docker 20.10.9 we started experience dropped connections to our Oracle database.

To reproduce I wrote a small .NET Framework application that opens a connection, makes a query, sleeps for 7 minutes and then tries to make another query. Running the same image on the default "nat" network does not exhibit this behavior.

### Reproduce

1. docker stack deploy -c stack.yml

``` stack.yml
version: "3.8"

services:
  odac:
    image: testodac:latest
```

2. Monitor logs for dropped connection.

### docker version

```bash
Client: Mirantis Container Runtime
 Version:           20.10.9
 API version:       1.41
 Go version:        go1.16.12m2
 Git commit:        591094d
 Built:             12/21/2021 21:34:30
 OS/Arch:           windows/amd64
 Context:           default
 Experimental:      true

Server: Mirantis Container Runtime
 Engine:
  Version:          20.10.9
  API version:      1.41 (minimum version 1.24)
  Go version:       go1.16.12m2
  Git commit:       9b96ce992b
  Built:            12/21/2021 21:33:06
  OS/Arch:          windows/amd64
  Experimental:     false
```

### docker info

```bash
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  cluster: Manage Mirantis Container Cloud clusters (Mirantis Inc., v1.9.0)
  registry: Manage Docker registries (Docker Inc., 0.1.0)

Server:
 Containers: 1
  Running: 1
  Paused: 0
  Stopped: 0
 Images: 3
 Server Version: 20.10.9
 Storage Driver: windowsfilter
  Windows:
 Logging Driver: json-file
 Plugins:
  Volume: local
  Network: ics internal l2bridge l2tunnel nat null overlay private transparent
  Log: awslogs etwlogs fluentd gcplogs gelf json-file local logentries splunk syslog
 Swarm: active
  NodeID: icbod14si9nowc2q0noek60el
  Is Manager: true
  ClusterID: t7tix610jmx0a6vau5kys3lxi
  Managers: 1
  Nodes: 1
  Default Address Pool: 10.0.0.0/8
  SubnetSize: 24
  Data Path Port: 4789
  Orchestration:
   Task History Retention Limit: 5
  Raft:
   Snapshot Interval: 10000
   Number of Old Snapshots to Retain: 0
   Heartbeat Tick: 1
   Election Tick: 10
  Dispatcher:
   Heartbeat Period: 5 seconds
  CA Configuration:
   Expiry Duration: 3 months
   Force Rotate: 0
  Autolock Managers: false
  Root Rotation In Progress: false
  Node Address: 10.135.28.243
  Manager Addresses:
   10.135.28.243:2377
 Default Isolation: process
 Kernel Version: 10.0 17763 (17763.1.amd64fre.rs5_release.180914-1434)
 Operating System: Windows Server 2019 Datacenter Version 1809 (OS Build 17763.3287)
 OSType: windows
 Architecture: x86_64
 CPUs: 2
 Total Memory: 8GiB
 Name: EC2AMAZ-5HKAR00
 ID: 6VW2:47K4:XZRJ:6S4L:UZHQ:MIIC:J5MJ:MWG5:FOWZ:NR6T:7MF6:IMNX
 Docker Root Dir: C:\ProgramData\docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by thaJeztah** [comment]
@evol262 @corhere ptal

**Comment by dmduggan** [comment]
I also reproduced this on Window Server 2022. So works on Server 2016, fails on Server 2019 and Server 2022.

**Comment by olljanat** [comment]
> To reproduce I wrote a small .NET Framework application that opens a connection, makes a query, sleeps for 7 minutes and then tries to make another query.

Can you share source code of that test app?

> Running the same image on the default "nat" network does not exhibit this behavior.

Where your DB server is hosted? Outside of swarm I assume?

> I also reproduced this on Window Server 2022. 

Single node swarm? All Windows Updates installed? How you install Docker to there?

**Comment by dmduggan** [comment]
Yes, it is a single node swarm with Oracle database outside of the swarm in AWS RDS.

Yes, I tested on Server 2022 after installing Windows Updates. I can try again on Server 2019.

I used the instructions/script  here to install:
https://docs.mirantis.com/mcr/20.10/install/mcr-windows.html

Test App Source Code:
[TestOdacCore.zip](https://github.com/moby/moby/files/9601299/TestOdacCore.zip)

**Comment by dmduggan** [comment]
Yes, we have the timeout set to 180, but works OK on Server 2016. Here is an example `connectionString` that we are using:
```Data Source=(DESCRIPTION=(ADDRESS_LIST=(ADDRESS=(PROTOCOL=TCP)(HOST=<HOST>)(PORT=1522)))(CONNECT_DATA=(SERVER=DEDICATED)(SERVICE_NAME=XE)));User id=<USER>;Password=<PASS>;Min Pool Size=2;Max Pool Size=5;Connection Timeout=180```

The join was just something I did to make debugging easier in Visual Studio. When I run it in a container it doesn't need to join anything.

**Comment by corhere** [comment]
The mechanism for how Windows containers attached to an overlay network access external services has changed in Windows 2019:
> [On Windows Server 2019 and above, overlay networks created by Docker Swarm leverage VFP NAT rules for outbound connectivity.](https://learn.microsoft.com/en-us/virtualization/windowscontainers/container-networking/network-drivers-topologies#:~:text=On%20Windows%20Server%202019%20and%20above%2C%20overlay%20networks%20created%20by%20Docker%20Swarm%20leverage%20VFP%20NAT%20rules%20for%20outbound%20connectivity.)

The VFP NAT rules have a default idle-connection timeout of 240 seconds. Windows will "forget" any connections from a container to an external host if no packets are sent or received for that duration, leaving a connection which appears to be open but does not actually route anywhere. The timeout for the `nat` network, and overlay NAT on Windows Server 2016, is 1800 seconds _[citation needed]_, which would explain why connections are not dropped in those configurations.

**Comment by olljanat** [comment]
> The VFP NAT rules have a default idle-connection timeout of 240 seconds.

@corhere That is good to know. Is there way to increase that value?

> Test App Source Code: [TestOdacCore.zip](https://github.com/moby/moby/files/9601299/TestOdacCore.zip)

@dmduggan If you can update code of real application then I would suggest to [ASP.NET health check](https://learn.microsoft.com/en-us/aspnet/core/host-and-deploy/health-checks?view=aspnetcore-6.0) (if that is ASP.NET, if not then using something similar in app logic) to application and configure Docker to polling it.

Then it will keep connection open and additionally Docker will restart container if connection stops working.

**Comment by corhere** [comment]
> Is there way to increase that value?

Not through Microsoft-sanctioned mechanisms, as far as I am aware. I have followed up in https://github.com/microsoft/Windows-Containers/issues/269#issuecomment-1252948256 asking for them to add one. In the meantime, I suggest experimenting with TCP Keep-alives to keep long lived connections from getting timed out.

**Comment by dmduggan** [comment]
The problem we saw with our real app is that since we are using a connection pool it seems like some of the connections are being "forgotten" when the pool is not fully utilized. So we would see intermittent failures. I'm guessing this may be more of a problem in our test environments than it would be in production. We had looked at keep-alives (`SQLNET.EXPIRE_TIME`, `Validate Connection=true`) but Oracle support discouraged us from doing that and instead of determine what in the networking was dropping the connections. But sounds like we found the source of that and kind of sounds like keep-alives may be our only option. Thank you all of the help and information.

**Comment by dmduggan** [comment]
I don't think the DB driver is "forgetting" the connection. It just that the driver doesn't recognize that the host OS has dropped the connection. When the host OS drops the connection we do see the connection disappear from the DB side but the next time the DB driver tries to use that connection from the pool we get the exception.
